metaflow/runtime.py

```diff
@@ -798,56 +836,110 @@ def __init__(
+                # on later resume attempt, we should check if the resume task is complete
+                # or not. This is to fix the issue where the resume leader was killed
+                # unexpectedly during cloning and never mark task complete.
+                try:
+                    task_completed = self.results["_task_ok"]
+                except DataException as e:
+                    pass
             else:
-                self._wait_for_clone = False
                 self._get_task_id(task_id)
 
             # Store the mapping from current_pathspec -> origin_pathspec which
             # will be useful for looking up origin_ds_set in find_origin_task.
             self.clone_pathspec_mapping[self._path] = origin.pathspec
             if self.step == "_parameters":
-                # We don't put _parameters on the queue so we either clone it or wait
-                # for it.
-                if not self._wait_for_clone:
+                # In the _parameters task, we need to resolve who is the resume leader.
+                self._is_resume_leader = False
+                resume_leader = None
+
+                if task_id_exists_already:
+                    # If the task id already exists, we need to check if current task is the resume leader in previous attempt.
+                    ds = self._flow_datastore.get_task_datastore(
+                        self.run_id, self.step, self.task_id
+                    )
+                    if not ds["_task_ok"]:
+                        raise MetaflowInternalError(
+                            "Externally cloned _parameters task did not succeed"
+                        )
+
+                    # Check if we should be the resume leader (maybe from previous attempt)
+                    if ds.has_metadata("_resume_leader", add_attempt=False):
+                        resume_leader = ds.load_metadata(
+                            ["_resume_leader"], add_attempt=False
+                        )["_resume_leader"]
+                        self._is_resume_leader = resume_leader == resume_identifier
+                else:
+                    # If the task id does not exist, current task is the resume leader.
+                    resume_leader = resume_identifier
+                    self._is_resume_leader = True
+
+                self.log(
+                    "Resume leader is %s." % resume_leader,
+                    system_msg=True,
+                )
+
+                if self._is_resume_leader:
+                    self.log(
+                        "Selected as the reentrant clone leader.",
+                        system_msg=True,
+                    )
                     # Clone in place without relying on run_queue.
                     self.new_attempt()
                     self._ds.clone(origin)
+                    # Set the resume leader be the task that calls the resume (first task to clone _parameters task).
+                    if resume_identifier:
```

Comment:
What is this check for? I thought the check should be more like: "if we actually created the task ID, we know we are the "first elected" resume leader so we write our identifier.

Also, shouldn't we actually error out if resume_identifier is not set in this case?

Reply:
this is not within the reentrant section, so it could still be empty, I feel that overwriting the same value twice should be fine so I did not check if we are "first elected" or first attempt (feel that checking this scenario makes the logic more complex and this operation is not too expensive).

Reply:
added more comments